Simultaneously support Wifi and Bluetooth wixels (updated)

jamorham opened the pull request:
Patch allows for running both Wifi and Bluetooth modes simultaneously. Avoids having to remember to switch modes when leaving the house etc. Useful for anyone who wants seamless coverage inside and out.
- Adds a new preference option for both modes and methods for starting both collection types.
- UI fixes to prevent double display of data 
- Duplicate reading checking now ignores two readings within 100 seconds (instead of previous 10) to ensure duplicates are not possible in this mode even with potential latency delays. (parakeet etc)
- Updated for current master branch as requested by @AdrianLxM 
- Minimal changes by mirroring existing coding style and application structure.

@connishiva

jamorham committed:
Sync with master

jamorham committed:
Add support for Bluetooth and Wifi Wixel modes simultaneously - backported from beta2 branch

jamorham committed:
Master merge

jamorham committed:
Add support for new collection service

AdrianLxM commented:
I just looked over the diff and it was basically the same as the one I was fine with 2.5 month ago.
I'd still like to test it but won't have the time until the end of the next weekend.

As a few people are using this it it good to get this merged it. -> Less forks with merge conflicts when we do bigger changes on master.

tzachi-dar commented:
Please let me look at it tonight...

mgranberry commented on app/src/main/java/com/eveningoutpost/dexdrip/utils/Preferences.java:
```diff
@@ -352,14 +359,24 @@ public boolean onPreferenceChange(Preference preference, Object newValue) {
 
-                    if(((String) newValue).compareTo("BluetoothWixel") != 0 && ((String) newValue).compareTo("DexcomShare") != 0 && ((String) newValue).compareTo("DexbridgeWixel") != 0) {
+                    if(((String) newValue).compareTo("BluetoothWixel") != 0 && ((String) newValue).compareTo("DexcomShare") != 0 && ((String) newValue).compareTo("DexbridgeWixel") != 0 && ((String) newValue).compareTo("WifiBlueToothWixel") != 0) {
+                 collectionCategory.removePreference(runInForeground);
```
Indentation is off here.

mgranberry replied:
There is a whitespace problem here too.

mgranberry replied:
The whitespace is off here.

mgranberry commented on app/src/main/java/com/eveningoutpost/dexdrip/Services/DexCollectionService.java:
```diff
@@ -100,7 +100,8 @@ public int onStartCommand(Intent intent, int flags, int startId) {
             return START_NOT_STICKY;
         }
-        if (CollectionServiceStarter.isBTWixel(getApplicationContext()) || CollectionServiceStarter.isDexbridgeWixel(getApplicationContext())) {
+        if (CollectionServiceStarter.isBTWixel(getApplicationContext()) || CollectionServiceStarter.isDexbridgeWixel(getApplicationContext())
```
I'm seeing this block of comparisons a lot.  Maybe there should be a CollectionServiceStarter.isWixel method to clean it and some of the other big ifs up?  Maybe it should just be that isBTWixel and isWifiWixel return true for the combined case.  It might not be in the scope of this PR, but this is starting to smell like we need some better abstractions around devices to make this easier.

mgranberry commented on app/src/main/java/com/eveningoutpost/dexdrip/Services/WifiCollectionService.java:
```diff
@@ -55,7 +55,8 @@ public int onStartCommand(Intent intent, int flags, int startId) {
         }
-        if (CollectionServiceStarter.isWifiWixel(getApplicationContext()) ) {
+        if (CollectionServiceStarter.isWifiWixel(getApplicationContext()) || CollectionServiceStarter.isWifiandBTWixel(getApplicationContext()) )
+        {
```
The '{' is no longer hanging.

mgranberry commented:
I'm seeing a lot of very long lines in the if statements.  Would you mind breaking them up onto multiple lines to make the code a little more readable?  I didn't see anything logically incorrect, but it seems like the functionality is split between wifi wixel and bt wixel functionality and there might not be a need for a third comparison.  As the conditionals get bigger there are more chances for errors and it becomes harder to add features.  It feels like this could probably be simplified by removing the isBtAndWifiWixel() test and modifying the two checks to return true for the combined case.  Am I missing something?

tzachi-dar commented on app/src/main/java/com/eveningoutpost/dexdrip/Models/TransmitterData.java:
```diff
@@ -58,7 +58,7 @@ public static TransmitterData create(byte[] buffer, int len, Long timestamp) {
+            if (lastTransmitterData != null && lastTransmitterData.raw_data == Integer.parseInt(data[0]) && Math.abs(lastTransmitterData.timestamp - timestamp) < (100000)) { //Stop allowing duplicate data, its bad! // jamorham: increased to 100 seconds for additional safety with multiple collection modes operating at the same time
                 return null;
             }
             if (data.length > 1) { transmitterData.sensor_battery_level = Integer.parseInt(data[1]); }
```
stupid thing, does not let me put the comment also from below.
need to increase the 10000, to 10000 also on line 77.
I would actually make this 120,000 to be more consistent.

tzachi-dar commented:
As for the very long ifs:
there are 3 of them all on the file DexCollectionService.
You can create one function that checks for all this cases, and replace them. will make the code more readable.

Last thing: in the file TransmiterData.java there is a try read - and if not found write. need to put that code (two places) under lock (then remove the random delay)

jamorham committed:
- merged changes from pull request #193
- broke up IF statements to multi-line for clarity
- auto-indented noted lines
- unified duplicate time period to 120s

jamorham committed:
randomDelay already removed

jamorham committed:
Merge remote-tracking branch 'upstream/master' into jamorham-btandwifi

Conflicts:
	app/src/main/java/com/eveningoutpost/dexdrip/Services/DexCollectionService.java

jamorham committed:
fix merge conflict

jamorham committed:
Merge branch 'jamorham-btandwifi2' into jamorham-btandwifi3

jamorham committed:
Formatting and merge logic fix

jamorham commented:
I have merged in the recent changes to master from #193 and updated my patches to improve the clarity of the "if" statements and applied auto-indentation to the noted blocks. I also unified the duplicate time period to 120s.

I haven't abstracted the preference reading methods (isBtWixel() etc). I see that as an overall future enhancement with the preferences system and this patch is designed to bring the functionality and mirror the existing coding style.

Xbridge+Wifi wixel support is something which would be good for the future but I don't have much experience using xbridge, so it isn't something I have attempted or tested and I am far less familiar with that part of the app. This is probably also something which would be easier to add with some future overhaul of preferences beyond the scope of the current PR.

mgranberry commented:
Thanks.  It looks good to me.

tzachi-dar commented:
Well, it seems that the random sleep is still there and there are still no locks.
I'm fine with checking this in, and than I'll do another pr that fixes this.

AdrianLxM commented:
Finally got to test it. Sorry for keeping you waiting. I'll merge in a minute.
